Lib/_pyio.py

```diff
@@ -978,6 +978,14 @@ def tell(self):
             raise ValueError("tell on closed file")
         return self._pos
 
+    def peek(self, size=-1):
+        pos = self.tell()
+        if size == 0:
+            size = -1
+        b = self.read(size)
```

Comment:
Would it be possible to implement it without touching the position? This code is not thread safe. I don't know if it's supposed to be thread-safe. Maybe add a private read method which has an argument to decide to move the position or not.

Same remark for C code.

Reply:
Good point; I implemented this now so that the position is not changed. I factored out a `peek_bytes` function in the C version that does not advance the position.

It seems though that neither the C nor the Python version of BytesIO are supposed to be thread safe. (They don’t use locks as e.g. `BufferedReader` does.) So I would suggest making them so would be a task for a different PR.